Release v0.2.0 (#31)
google/JetStream: README.md

JetStream is a fast library for LLM inference and serving on TPUs (and GPUs in future -- PRs welcome).

## Documentation

- [Online Inference with MaxText on v5e Cloud TPU VM](#jetstream-maxtext-inference-on-v5e-cloud-tpu-vm-user-guide)
- [JetStream Standalone Local Setup](#jetstream-standalone-local-setup)

---

# JetStream MaxText Inference on v5e Cloud TPU VM User Guide

## Outline

1. Prerequisites: Prepare your GCP project and connect to Cloud TPU VM
2. Download the JetStream and MaxText github repository
3. Setup your MaxText JetStream environment
4. Convert Model Checkpoints
5. Run the JetStream MaxText server
6. Send a test request to the JetStream MaxText server
7. Run benchmarks with the JetStream MaxText server
8. Clean up

## Prerequisites: Prepare your GCP project and connect to Cloud TPU VM

Follow the steps in [Manage TPU resources | Google Cloud](https://cloud.google.com/tpu/docs/managing-tpus-tpu-vm) to create a Cloud TPU VM (Recommend TPU type: `v5litepod-8`) and connect to the Cloud TPU VM.

## Step 1: Download JetStream and the MaxText github repository

```bash

> git clone -b jetstream-v0.2.0 https://github.com/google/maxtext.git

Reviewer: Will this need to be updated with every jetstream release?

Author: For this user guide demo, we want to use the stable release version. Later we can update it or keep it as it is.

Reviewer: Got it, thanks!

git clone -b v0.2.0 https://github.com/google/JetStream.git
```

## Step 2: Setup MaxText

```bash
# Create a python virtual environment for the demo.
sudo apt install python3.10-venv
python -m venv .env
source .env/bin/activate

# Setup MaxText.
cd maxtext/
bash setup.sh
```

## Step 3: Convert Model Checkpoints 

You can run the JetStream MaxText Server with Gemma and Llama2 models. This section describes how to run the JetStream MaxText server with various sizes of these models.

### Use a Gemma model checkpoint

*   You can download a [Gemma checkpoint from Kaggle](https://www.kaggle.com/models/google/gemma/frameworks/maxText/variations/7b). 
*   After downloading checkpoints, copy them to your GCS bucket at `$CHKPT_BUCKET`.
    *   `gsutil -m cp -r ${YOUR_CKPT_PATH} ${CHKPT_BUCKET}`
    *   Please refer to the [conversion script](https://github.com/google/JetStream/blob/main/jetstream/tools/maxtext/model_ckpt_conversion.sh) for an example of `$CHKPT_BUCKET`.
*   Then, using the following command to convert the Gemma checkpoint into a MaxText compatible unscanned checkpoint.

```bash

> # bash ../JetStream/jetstream/tools/maxtext/model_ckpt_conversion.sh ${MODEL} ${MODEL_VARIATION} ${CHKPT_BUCKET}
> bash ../JetStream/jetstream/tools/maxtext/model_ckpt_conversion.sh gemma 7b ${CHKPT_BUCKET}

Reviewer: ```suggestion
# bash ../JetStream/jetstream/tools/maxtext/model_ckpt_conversion.sh ${MODEL} ${MODEL_VARIATION} ${CHKPT_BUCKET}

# For gemma-7b
bash ../JetStream/jetstream/tools/maxtext/model_ckpt_conversion.sh gemma 7b ${CHKPT_BUCKET}
```

*   After downloading checkpoints, copy them to your GCS bucket at `$CHKPT_BUCKET`.
    *   `gsutil -m cp -r ${YOUR_CKPT_PATH} ${CHKPT_BUCKET}`
    *   Please refer to the [conversion script](https://github.com/google/JetStream/blob/main/jetstream/tools/maxtext/model_ckpt_conversion.sh) for an example of `$CHKPT_BUCKET`.
*   Then, using the following command to convert the Llama2 checkpoint into a MaxText compatible unscanned checkpoint.

```bash
# bash ../JetStream/jetstream/tools/maxtext/model_ckpt_conversion.sh ${MODEL} ${MODEL_VARIATION} ${CHKPT_BUCKET}

# For llama2-7b
bash ../JetStream/jetstream/tools/maxtext/model_ckpt_conversion.sh llama2 7b ${CHKPT_BUCKET}

# For llama2-13b
bash ../JetStream/jetstream/tools/maxtext/model_ckpt_conversion.sh llama2 13b ${CHKPT_BUCKET}
```

Note: For more information about the Llama2 model and checkpoints, see [About Llama2](https://github.com/google/maxtext/blob/main/getting_started/Run_Llama2.md).

## Step4: Run the JetStream MaxText server

### Create model config environment variables for server flags

You can export the following environment variables based on the model you used.

> *   You can copy and export the `UNSCANNED_CKPT_PATH` from the model\_ckpt\_conversion.sh output.

Reviewer: ```suggestion
*   You can copy and export the `UNSCANNED_CKPT_PATH` from the `model\_ckpt\_conversion.sh` output.
```
```suggestion
*   You can copy and export the `UNSCANNED_CKPT_PATH` from the model\_ckpt\_conversion.sh output.
```

> *   You can copy and export the `UNSCANNED_CKPT_PATH` from the model\_ckpt\_conversion.sh output.

Reviewer: Can you clarify this step? i.e. `UNSCANNED_CKPT_PATH` is a variable exported from `model_ckpt_conversion.sh`, which points to the unscanned checkpoint that it generated. 

And should we also explain that: Unscanned checkpoints are more efficient decoding, whereas scanned are better for training.

*   Configure the [flags](#jetstream-maxtext-server-flag-descriptions) passing into the JetStream MaxText server

```bash
export TOKENIZER_PATH=assets/tokenizer.llama2
export LOAD_PARAMETERS_PATH=${UNSCANNED_CKPT_PATH}
export MAX_PREFILL_PREDICT_LENGTH=1024
export MAX_TARGET_LENGTH=2048
export MODEL_NAME=llama2-7b
export ICI_FSDP_PARALLELISM=1
export ICI_AUTOREGRESSIVE_PARALLELISM=-1
export ICI_TENSOR_PARALLELISM=1
export SCAN_LAYERS=false
export WEIGHT_DTYPE=bfloat16
export PER_DEVICE_BATCH_SIZE=6
```

#### Create Llama2-13b environment variables for server flags

*   Configure the [flags](#jetstream-maxtext-server-flag-descriptions) passing into the JetStream MaxText server

```bash
export TOKENIZER_PATH=assets/tokenizer.llama2
export LOAD_PARAMETERS_PATH=${UNSCANNED_CKPT_PATH}
export MAX_PREFILL_PREDICT_LENGTH=1024
export MAX_TARGET_LENGTH=2048
export MODEL_NAME=llama2-13b
export ICI_FSDP_PARALLELISM=1
export ICI_AUTOREGRESSIVE_PARALLELISM=-1
export ICI_TENSOR_PARALLELISM=1
export SCAN_LAYERS=false
export WEIGHT_DTYPE=bfloat16
export PER_DEVICE_BATCH_SIZE=2
```

### Run the following command to start the JetStream MaxText server

```bash
cd ~/maxtext

> python MaxText/maxengine_server.py \
> MaxText/configs/base.yml \
> tokenizer_path=${TOKENIZER_PATH} \
> load_parameters_path=${LOAD_PARAMETERS_PATH} \
> max_prefill_predict_length=${MAX_PREFILL_PREDICT_LENGTH} \
> max_target_length=${MAX_TARGET_LENGTH} \
> model_name=${MODEL_NAME} \
> ici_fsdp_parallelism=${ICI_FSDP_PARALLELISM} \
> ici_autoregressive_parallelism=${ICI_AUTOREGRESSIVE_PARALLELISM} \
> ici_tensor_parallelism=${ICI_TENSOR_PARALLELISM} \
> scan_layers=${SCAN_LAYERS} \
> weight_dtype=${WEIGHT_DTYPE} \
> per_device_batch_size=${PER_DEVICE_BATCH_SIZE}

Reviewer: ```suggestion
python MaxText/maxengine_server.py \
  MaxText/configs/base.yml \
  tokenizer_path=${TOKENIZER_PATH} \
  load_parameters_path=${LOAD_PARAMETERS_PATH} \
  max_prefill_predict_length=${MAX_PREFILL_PREDICT_LENGTH} \
  max_target_length=${MAX_TARGET_LENGTH} \
  model_name=${MODEL_NAME} \
  ici_fsdp_parallelism=${ICI_FSDP_PARALLELISM} \
  ici_autoregressive_parallelism=${ICI_AUTOREGRESSIVE_PARALLELISM} \
  ici_tensor_parallelism=${ICI_TENSOR_PARALLELISM} \
  scan_layers=${SCAN_LAYERS} \
  weight_dtype=${WEIGHT_DTYPE} \
  per_device_batch_size=${PER_DEVICE_BATCH_SIZE}
```

```

### JetStream MaxText Server flag descriptions:

*   tokenizer\_path: file path to a tokenizer (should match your model)

> *   load\_parameters\_path: Loads a just parameters from a specific directory

Reviewer: ```suggestion
*   load\_parameters\_path: Loads the parameters (no optimizer states) from a specific directory
```

```bash
# Enable int8 quantization for both weights and KV cache
export QUANTIZATION=int8
export QUANTIZE_KVCACHE=true

# For Gemma 7b model, change per_device_batch_size to 12 to optimize performance. 
export PER_DEVICE_BATCH_SIZE=12

cd ~/maxtext
python MaxText/maxengine_server.py \
MaxText/configs/base.yml \
tokenizer_path=${TOKENIZER_PATH} \
load_parameters_path=${LOAD_PARAMETERS_PATH} \
max_prefill_predict_length=${MAX_PREFILL_PREDICT_LENGTH} \
max_target_length=${MAX_TARGET_LENGTH} \
model_name=${MODEL_NAME} \
ici_fsdp_parallelism=${ICI_FSDP_PARALLELISM} \
ici_autoregressive_parallelism=${ICI_AUTOREGRESSIVE_PARALLELISM} \
ici_tensor_parallelism=${ICI_TENSOR_PARALLELISM} \
scan_layers=${SCAN_LAYERS} \
weight_dtype=${WEIGHT_DTYPE} \
per_device_batch_size=${PER_DEVICE_BATCH_SIZE} \
quantization=${QUANTIZATION} \
quantize_kvcache=${QUANTIZE_KVCACHE}
```

### Benchmarking Gemma-7b

> ```bash
> # For Gemma 7b, plz use Gemma tokenizer (must use a tokenizer that matches your model; double check if you are using tokenizer.gemma) and ShareGPT dataset; add `--warmup-first` flag for your 1st run to warmup the server.

Reviewer: ```suggestion
Instructions
- Download the ShareGPT dataset
- Make sure to use the Gemma tokenizer (tokenizer.gemma) when running Gemma 7b.
- Add `--warmup-first` flag for your 1st run to warmup the server

```bash
```

# Activate the python virtual environment we created in Step 2.
cd ~
source .env/bin/activate

# download dataset
wget https://huggingface.co/datasets/anon8231489123/ShareGPT_Vicuna_unfiltered/resolve/main/ShareGPT_V3_unfiltered_cleaned_split.json

# run benchmark with the downloaded dataset and the tokenizer in maxtext

> # You can control the qps by setting `--request-rate`, the default value is inf.

Reviewer: What is the effect of `--request-rate` being set to inf?

Author: dispatch all requests at once.

python JetStream/benchmarks/benchmark_serving.py \
--tokenizer /home/$USER/maxtext/assets/tokenizer.gemma \ 
--num-prompts 1000 \
--dataset sharegpt \
--dataset-path ~/ShareGPT_V3_unfiltered_cleaned_split.json \
--max-output-length 1024 \
--request-rate 5 \
--warmup-first true
```

### Benchmarking Llama2-\*b

```bash

> # Same as Gemma-7b except for tokenizer (must use a tokenizer that matches your model). 

Reviewer: ```suggestion
# Same as Gemma-7b except for the tokenizer (must use a tokenizer that matches your model, which should now be tokenizer.llama2). 
```
